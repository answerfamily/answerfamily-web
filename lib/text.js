import React from 'react';

const BREAK = { $$BREAK: true };

/**
 * Invokes traverseForStrings for each item in elems.
 * When `BREAK` is received, break traversal immediately.
 *
 * @param {*} elem Array of elements to traverse
 * @param {Function} callback passed to traverseForStrings()
 */
function traverseElems(elems, callback) {
  const result = [];
  for (let i = 0; i < elems.length; i += 1) {
    const returnValue = traverseForStrings(elems[i], callback);
    if (returnValue === BREAK) break;
    result.push(returnValue);
  }

  return result;
}

/**
 * Traverses elem tree for strings, returns callback(string)
 * @param {*} elem
 * @param {Function} callback
 */
function traverseForStrings(elem, callback) {
  switch (true) {
    case typeof elem === 'string':
      return callback(elem);

    case elem instanceof Array:
      return traverseElems(elem, callback);

    case React.isValidElement(elem): {
      const children = React.Children.toArray(elem.props.children);
      const newChildren = traverseElems(children, callback);

      // No need to clone element if new children is identical with the original
      //
      if (
        children.length === newChildren.length &&
        children.every((child, idx) => child === newChildren[idx])
      ) {
        return elem;
      }

      return React.cloneElement(elem, {}, newChildren);
    }
    default:
      return null;
  }
}

function shortenUrl(s, maxLength) {
  try {
    s = decodeURIComponent(s);
  } catch (e) {
    // Probably malformed URI components.
    // Do nothing, just use original string
  }
  return s.length <= maxLength
    ? s
    : `${s.slice(0, maxLength / 2)}â‹¯${s.slice(-maxLength / 2)}`;
}

function flatternPureStrings(tokens) {
  return tokens.every(token => typeof token === 'string')
    ? tokens.join()
    : tokens;
}

const urlRegExp = /(https?:\/\/\S+)/;

/**
 * Wrap <a> around hyperlinks inside a react element or string.
 *
 * @param {*} elem React element, string, array of string & react elements
 * @param {<maxLength: Number, blank: Boolean>} options
 */
export function linkify(elem, { maxLength = 80, props = {} } = {}) {
  return traverseForStrings(elem, str => {
    if (!str) return str;

    const tokenized = str.split(urlRegExp).map((s, i) =>
      s.match(urlRegExp) ? (
        <a key={`link${i}`} href={s} {...props}>
          {shortenUrl(s, maxLength)}
        </a>
      ) : (
        s
      )
    );

    return flatternPureStrings(tokenized);
  });
}

const newLinePattern = '(\r\n|\r|\n)';
// Spaces around new line pattern should be safe to trim, because we are placing <br>
// on the newLinePattern.
const newLineRegExp = RegExp(` *${newLinePattern} *`, 'g');

/**
 * Place <br> for each line break.
 * Automatically trims away leading & trailing line breaks.
 *
 * @param {*} elem React element, string, array of string & react elements
 */
export function nl2br(elem) {
  return traverseForStrings(elem, str => {
    if (!str) return str;

    const tokenized = str
      .split(newLineRegExp)
      .filter(token => token !== '') // Filter out empty strings
      .map((line, idx) =>
        line.match(newLineRegExp) ? <br key={`br${idx}`} /> : line
      );

    // If the tokenized contains only string, join into one single string.
    //
    return flatternPureStrings(tokenized);
  });
}

/**
 * Truncates the given elem to the specified wordCount.
 * When truncated, appends moreElem to the end of the string.
 *
 * @param {*} elem React element, string, array of string & react elements
 * @param {<wordCount: Number, moreElem: ReactElement>} options
 */
export function truncate(elem, { wordCount = Infinity, moreElem = null } = {}) {
  let currentWordCount = 0;
  const result = traverseForStrings(elem, str => {
    if (currentWordCount >= wordCount) return BREAK;

    currentWordCount += str.length;
    const exceededCount = currentWordCount - wordCount;

    return exceededCount <= 0 ? str : str.slice(0, -exceededCount);
  });

  switch (true) {
    // Not exceeding wordCount, just return the original element
    case currentWordCount <= wordCount:
      return elem;

    // If the result is an array, append moreElem
    case result instanceof Array:
      return result.concat(moreElem);

    // Others, including result being a string or React Element.
    default:
      return [result, moreElem];
  }
}
