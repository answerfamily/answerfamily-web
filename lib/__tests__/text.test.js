import React from 'react';
import { linkify, nl2br, truncate } from '../text';

describe('linkify', () => {
  it('does nothing on strings, arrays and elements without links', () => {
    expect(linkify('foo')).toBe('foo');

    const singleLevelElem = <p>foo</p>;
    expect(linkify(singleLevelElem)).toBe(singleLevelElem);

    const nestedElem = (
      <p>
        foo <span> bar </span> foo2
      </p>
    );
    expect(linkify(nestedElem)).toBe(nestedElem);

    const array = ['foo', 'bar', <span key="elem">lala</span>];
    expect(linkify(array)).toEqual(array);
  });

  it('attach link on strings containing links', () => {
    expect(linkify('Please go to http://google.com')).toMatchSnapshot();
  });

  it('attach link on simple elements containing links', () => {
    expect(
      linkify(<span>Please go to http://google.com</span>)
    ).toMatchSnapshot();
  });

  it('attach link on nested elements containing links', () => {
    expect(
      linkify(
        <span>
          Please go to <span>http://google.com</span> lalala
        </span>
      )
    ).toMatchSnapshot();
  });

  it('attach link on arrays containing elements that has link', () => {
    expect(
      linkify([
        'Please go to http://google.com',
        <span key="elem">Please go to http://google.com</span>,
      ])
    ).toMatchSnapshot();
  });

  it('decodes URI encoded URLs', () => {
    expect(
      linkify(
        'Please go to http://www.rumtoast.com/5444/line%E7%BE%A4%E7%B5%84%E8%A1%8C%E5%8B%95%E6%A2%9D%E7%A2%BC%E9%82%80%E8%AB%8B-%E4%B8%80%E5%AE%9A%E8%A6%81%E9%97%9C%E6%8E%89%EF%BC%8C%E4%B8%8D%E7%84%B6%E9%A7%AD%E5%AE%A2%E6%9C%83%E5%85%A5%E4%BE%B5'
      )
    ).toMatchSnapshot();
  });

  it('decodes URLS that is too long', () => {
    expect(
      linkify(
        'Please go to http://www.rumtoast.com/5444/line%E7%BE%A4%E7%B5%84%E8%A1%8C%E5%8B%95%E6%A2%9D%E7%A2%BC%E9%82%80%E8%AB%8B-%E4%B8%80%E5%AE%9A%E8%A6%81%E9%97%9C%E6%8E%89%EF%BC%8C%E4%B8%8D%E7%84%B6%E9%A7%AD%E5%AE%A2%E6%9C%83%E5%85%A5%E4%BE%B5',
        { maxLength: 10 }
      )
    ).toMatchSnapshot();
  });

  it('sets prop to link', () => {
    expect(
      linkify('Please go to http://google.com', {
        props: { target: '_blank' },
      })
    ).toMatchSnapshot();
  });
});

describe('nl2br', () => {
  it('does nothing on strings, arrays and elements without line breaks', () => {
    expect(nl2br('foo')).toBe('foo');

    const singleLevelElem = <p>foo</p>;
    expect(nl2br(singleLevelElem)).toBe(singleLevelElem);

    const nestedElem = (
      <p>
        foo <span> bar </span> foo2
      </p>
    );
    expect(nl2br(nestedElem)).toBe(nestedElem);

    const array = ['foo', 'bar', <span key="elem">lala</span>];
    expect(nl2br(array)).toEqual(array);
  });

  it('inserts <br> on line breaks in strings, with ending <br>s trimmed', () => {
    expect(nl2br('Foo\nBar')).toMatchSnapshot();

    expect(
      nl2br(`This should be first line with empty next line
          This should be second line with no <br> afterwards`)
    ).toMatchSnapshot();
  });

  it('inserts <br> on line breaks within elements', () => {
    expect(
      nl2br(
        <p>
          {`This should be first line
              This should be second line with no <br> afterwards`}
        </p>
      )
    ).toMatchSnapshot();
  });

  it('preserves line ends on the end of tag', () => {
    expect(
      nl2br([
        <a href="" key="link">
          http://www.appledaily.com.tw/realtimenews/article/new/20170817/1184132/
        </a>,
        '\n15少年集體性侵驢子　全染上狂犬病',
      ])
    ).toMatchSnapshot();
  });
});

describe('truncate', () => {
  it('does nothing if wordCount is infinity', () => {
    expect(truncate('12345')).toBe('12345');

    const singleLevelElem = <p>This is not truncated</p>;
    expect(singleLevelElem).toBe(singleLevelElem);
  });

  it('truncates strings', () => {
    const moreElem = <button key="btn">...more</button>;

    expect(truncate('12345', { moreElem, wordCount: 3 })).toEqual([
      '123',
      moreElem,
    ]);
  });

  it('truncates nested DOM', () => {
    const moreElem = <button key="btn">...more</button>;
    const inputElem = (
      <div>
        12
        <p>
          34
          <a href="">56</a>
          78
          <a href="">90</a>
          12
        </p>
        34
      </div>
    );

    expect(truncate(inputElem, { moreElem, wordCount: 3 })).toMatchSnapshot();
    expect(truncate(inputElem, { moreElem, wordCount: 5 })).toMatchSnapshot();
    expect(truncate(inputElem, { moreElem, wordCount: 7 })).toMatchSnapshot();
    expect(truncate(inputElem, { moreElem, wordCount: 9 })).toMatchSnapshot();
  });
});
