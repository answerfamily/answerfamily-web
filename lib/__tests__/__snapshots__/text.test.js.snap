// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`linkify attach link on arrays containing elements that has link 1`] = `
Array [
  Array [
    "Please go to ",
    <a
      href="http://google.com"
    >
      http://google.com
    </a>,
    "",
  ],
  <span>
    Please go to 
    <a
      href="http://google.com"
    >
      http://google.com
    </a>
    
  </span>,
]
`;

exports[`linkify attach link on nested elements containing links 1`] = `
<span>
  Please go to 
  <span>
    
    <a
      href="http://google.com"
    >
      http://google.com
    </a>
    
  </span>
   lalala
</span>
`;

exports[`linkify attach link on simple elements containing links 1`] = `
<span>
  Please go to 
  <a
    href="http://google.com"
  >
    http://google.com
  </a>
  
</span>
`;

exports[`linkify attach link on strings containing links 1`] = `
Array [
  "Please go to ",
  <a
    href="http://google.com"
  >
    http://google.com
  </a>,
  "",
]
`;

exports[`linkify decodes URI encoded URLs 1`] = `
Array [
  "Please go to ",
  <a
    href="http://www.rumtoast.com/5444/line%E7%BE%A4%E7%B5%84%E8%A1%8C%E5%8B%95%E6%A2%9D%E7%A2%BC%E9%82%80%E8%AB%8B-%E4%B8%80%E5%AE%9A%E8%A6%81%E9%97%9C%E6%8E%89%EF%BC%8C%E4%B8%8D%E7%84%B6%E9%A7%AD%E5%AE%A2%E6%9C%83%E5%85%A5%E4%BE%B5"
  >
    http://www.rumtoast.com/5444/line群組行動條碼邀請-一定要關掉，不然駭客會入侵
  </a>,
  "",
]
`;

exports[`linkify decodes URLS that is too long 1`] = `
Array [
  "Please go to ",
  <a
    href="http://www.rumtoast.com/5444/line%E7%BE%A4%E7%B5%84%E8%A1%8C%E5%8B%95%E6%A2%9D%E7%A2%BC%E9%82%80%E8%AB%8B-%E4%B8%80%E5%AE%9A%E8%A6%81%E9%97%9C%E6%8E%89%EF%BC%8C%E4%B8%8D%E7%84%B6%E9%A7%AD%E5%AE%A2%E6%9C%83%E5%85%A5%E4%BE%B5"
  >
    http:⋯駭客會入侵
  </a>,
  "",
]
`;

exports[`linkify sets prop to link 1`] = `
Array [
  "Please go to ",
  <a
    href="http://google.com"
    target="_blank"
  >
    http://google.com
  </a>,
  "",
]
`;

exports[`mark marks specified text 1`] = `
Array [
  "
        下學期國中教科書教性解放
        1.",
  <mark>
    國中如何性滿足
  </mark>,
  "
        2.如何",
  <mark>
    性愛自拍
  </mark>,
  "
        3.",
  <mark>
    鼓勵統計性伴侶人數
  </mark>,
  "
        4.",
  <mark>
    介紹師生戀
  </mark>,
  "
        5.",
  <mark>
    介紹人動物戀/交合
  </mark>,
  "！
        您看了除了生氣還能做甚麼？
      ",
]
`;

exports[`nl2br inserts <br> on line breaks in strings, with ending <br>s trimmed 1`] = `
Array [
  "Foo",
  <br />,
  "Bar",
]
`;

exports[`nl2br inserts <br> on line breaks in strings, with ending <br>s trimmed 2`] = `
Array [
  "This should be first line with empty next line",
  <br />,
  "This should be second line with no <br> afterwards",
]
`;

exports[`nl2br inserts <br> on line breaks within elements 1`] = `
<p>
  This should be first line
  <br />
  This should be second line with no &lt;br&gt; afterwards
</p>
`;

exports[`nl2br preserves line ends on the end of tag 1`] = `
Array [
  <a
    href=""
  >
    http://www.appledaily.com.tw/realtimenews/article/new/20170817/1184132/
  </a>,
  Array [
    <br />,
    "15少年集體性侵驢子　全染上狂犬病",
  ],
]
`;

exports[`truncate truncates nested DOM 1`] = `
Array [
  <div>
    12
    <p>
      3
      <a
        href=""
      />
    </p>
  </div>,
  <button>
    ...more
  </button>,
]
`;

exports[`truncate truncates nested DOM 2`] = `
Array [
  <div>
    12
    <p>
      34
      <a
        href=""
      >
        5
      </a>
    </p>
  </div>,
  <button>
    ...more
  </button>,
]
`;

exports[`truncate truncates nested DOM 3`] = `
Array [
  <div>
    12
    <p>
      34
      <a
        href=""
      >
        56
      </a>
      7
      <a
        href=""
      />
    </p>
  </div>,
  <button>
    ...more
  </button>,
]
`;

exports[`truncate truncates nested DOM 4`] = `
Array [
  <div>
    12
    <p>
      34
      <a
        href=""
      >
        56
      </a>
      78
      <a
        href=""
      >
        9
      </a>
    </p>
  </div>,
  <button>
    ...more
  </button>,
]
`;
